source/server/listener_impl.cc

```diff
@@ -461,6 +498,59 @@ void ListenerImpl::buildFilterChains() {
   filter_chain_manager_.addFilterChain(config_.filter_chains(), builder, filter_chain_manager_);
 }
 
+void ListenerImpl::buildFakeFilterChains() {
+  Server::Configuration::TransportSocketFactoryContextImpl transport_factory_context(
+      parent_.server_.admin(), parent_.server_.sslContextManager(), listenerScope(),
+      parent_.server_.clusterManager(), parent_.server_.localInfo(), parent_.server_.dispatcher(),
+      parent_.server_.random(), parent_.server_.stats(), parent_.server_.singletonManager(),
+      parent_.server_.threadLocal(), validation_visitor_, parent_.server_.api());
+  transport_factory_context.setInitManager(*dynamic_init_manager_);
+  ListenerFilterChainFactoryBuilder builder(*this, transport_factory_context);
+  filter_chain_manager_.addFakeFilterChain(config_.filter_chains(), builder, filter_chain_manager_);
+}
+
+void ListenerImpl::buildRealFilterChains(
+    const envoy::config::listener::v3::FilterChain* const& filter_chain,
+    FilterChainRebuildInfoPtr rebuild_info) {
+  // auto filter_chain_name = filter_chain->name();
+
+  // Trigger rebuild request only on the arrival of the first filter_chain.
+  bool on_first_request = false;
+
+  // Find/create rebuilder for this filter chain message.
+  if (filter_chain_rebuilder_map_.find(filter_chain) == filter_chain_rebuilder_map_.end()) {
+    filter_chain_rebuilder_map_[filter_chain] =
+        std::make_unique<FilterChainRebuilder>(*this, filter_chain);
+    on_first_request = true;
+  }
+
+  const auto& rebuilder = std::move(filter_chain_rebuilder_map_[filter_chain]);
+  rebuilder->addRebuildInfo(std::move(rebuild_info));
+
+  if (rebuilder->isCompleted()) {
+    rebuilder->callbackToWorkers();
+  }
+
+  // Request dependencies only on the first time.
+  if (on_first_request) {
+    Server::Configuration::TransportSocketFactoryContextImpl transport_factory_context(
+        parent_.server_.admin(), parent_.server_.sslContextManager(), listenerScope(),
+        parent_.server_.clusterManager(), parent_.server_.localInfo(), parent_.server_.dispatcher(),
+        parent_.server_.random(), parent_.server_.stats(), parent_.server_.singletonManager(),
+        parent_.server_.threadLocal(), validation_visitor_, parent_.server_.api());
+    transport_factory_context.setInitManager(rebuilder->initManager());
+    ListenerFilterChainFactoryBuilder builder(*this, transport_factory_context);
+    filter_chain_manager_.addRealFilterChain(std::move(filter_chain), builder,
+                                             filter_chain_manager_);
+
+    rebuilder->startRebuilding();
```

Comment:
Given 'rebuildInfo' from workers, find if a rebuilder(for this filter chain message) exists. If not create one and only on the first time, go request dependencies and start rebuilding. The part of dependency request is the just the same as the current one.